Added new tests to wp-admin-order-edit.test.js for covering merchant management actions over downloadable products.

alopezari opened the pull request:
### All Submissions:

* [ x ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [ x ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ x ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Fixed a test from the `merchant/wp-admin-order-edit` spec that could return false positive results.

Updated previous `merchant/wp-admin-order-edit` tests so that orders are created through the API as a precondition, reducing the test run duration of the spec by ~8 seconds.

Updated the `can update settings` test from the `merchant/wp-admin-settings-product` spec so that it also covered setting the "X-Accel-Redirect/X-Sendfile" download method and enabling / disabling the option of adding a hash to the filename.  

Added the following tests to the `merchant/wp-admin-order-edit` spec:
- Can add downloadable product permissions to order without product.
- Can add downloadable product permissions to order with product.
- Can edit downloadable product permissions.
- Can revoke downloadable product permissions.
- Should not allow downloading a product if download attempts are exceeded.
- Should not allow downloading a product if expiration date is exceeded.

In order to improve tests structure and readability, and letting developers be able to reuse some merchants actions such as adding/updating/revoking a downloadable product permission to an order, the following functions have been added to `merchant.js`:
- addDownloadableProductPermission( productName)
- updateDownloadableProductPermission( productName, expirationDate, downloadsRemaining )
- revokeDownloadableProductPermission ( productName )

For the same reason, the following functions were also added to `merchant.js`:
- openDownloadLink()
- verifyCannotDownloadFromBecause( page, reason ) => this one is read: verify cannot download from `page` because `reason`

Also, the following functions were added to `page-utils.js`:
- getSelectorAttribute ( selector, attribute ) => useful for retrieving HTML attributes from elements, like 'value', 'href', 'placeholder' and others
- verifyValueOfElementAttribute ( selector, attribute, expectedValue ) => makes use of the function above to verify that the attribute retrieved matches the expected value
- orderPageSaveChanges() => this action is performed multiple times in the `merchant/wp-admin-order-edit` spec, so it has been extracted to a function

Finally, a new wrapper of the createSimpleProduct() function has been added to `components.js` to create easily new downloadable products:
- createSimpleDownloadableProduct ( name, downloadLimit, downloadName, price )

**These new tests will increase the `Downloadable Products` coverage for merchants.**

### How to test the changes in this Pull Request:

1. Go to the main `woocommerce` repo folder.
2. Run the e2e tests: `pnpm nx test-e2e woocommerce`.

Alternative:
1. Go to `woocommerce/plugins/woocommerce`.
2. Run the `order-edit` tests: `pnpx wc-e2e test:e2e ./tests/e2e/specs/wp-admin/order-edit.test.js `

The alternative is faster, but the main suggestion is preferred because it would help finding potential regression issues in the test framework.

### Other information:

* [ x ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ x ] Have you written new tests for your changes, as applicable?
* [ x ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

See `Changes proposed in this Pull Request`.

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

alopezari committed:
Added new tests to wp-admin-order-edit.test.js for covering merchant management actions over downloadable products.

zhongruige commented:
Hi @alopezari! Thank you for getting this PR in, we appreciate it! Just one minor thing I wanted to note to also update on this PR--in addition to updating the README (thanks so much for doing that on this PR!) when we add new functionality or make changes to any of the packages we maintain, we also update the `CHANGELOG` with those changes as well:

* [e2e-core-tests changelog](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/e2e-core-tests/CHANGELOG.md)
* [e2e-utils](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/e2e-utils/CHANGELOG.md)

The changelog is divided into the following sections, depending on the changes:
* Fixed
* Added
* Changed

Let me know if you have any questions!

alopezari committed:
Updated changelogs from e2e-utils and e2e-core packages.

alopezari commented:
Thanks for the instructions @zhongruige! I've updated both changelogs and pushed the changes.

rodelgc commented on packages/js/e2e-core-tests/src/specs/merchant/wp-admin-order-edit.test.js:
```diff
@@ -69,14 +77,152 @@ const runEditOrderTest = () => {
+		});
+
+		afterAll( async () => {
+			await withRestApi.deleteAllOrders();
```
Hi @alopezari 👋 

We can probably delete this `afterAll` block because we already have one in [line 28](https://github.com/alopezari/woocommerce/blob/8561de2f064ba0bc60dde9c2cd144609d1dd507f/packages/js/e2e-core-tests/src/specs/merchant/wp-admin-order-edit.test.js#L28)?

alopezari replied:
Hi @rodelgc! Yes, I missed it, sorry. Thanks for spotting this! Will fix it later today.

rodelgc replied:
No worries! 😄

alopezari commented on packages/js/e2e-core-tests/src/specs/merchant/wp-admin-order-edit.test.js:
```diff
@@ -69,14 +77,152 @@ const runEditOrderTest = () => {
+	  		await merchant.verifyCannotDownloadFromBecause( downloadPage, expectedReason );
+		} );
+
+		afterEach( async () => {
+			await withRestApi.deleteAllOrders();
+			await withRestApi.deleteAllProducts();
+		} );
```
@rodelgc If you're happy with this, I will also move this `afterEach` to the line 112 so that both `describe` blocks follow the same structure.

rodelgc replied:
Regarding the location of the `afterEach` block, yep it would be nice to have it right after `beforeEach` although it isn't a strict requirement. Still, it's good to relocate it there for styling consistency.

Regarding deleting of all orders and products, I see why you went this approach. It's because some of the tests you added will create another order or product aside from the ones that `beforeEach` already created. I wonder if there's a more efficient way of cleaning them up without deleting all orders and products, and a way to set them up without creating duplicate orders/products in some of the new tests. I'm afraid that deleting all of them after each test would interfere with other tests when our test suite gained the ability to be run in parallel in the future.

alopezari replied:
> Regarding the location of the afterEach block, yep it would be nice to have it right after beforeEach although it isn't a strict requirement. Still, it's good to relocate it there for styling consistency.

Indeed! I'll relocate it.

> Regarding deleting of all orders and products, I see why you went this approach. It's because some of the tests you added will create another order or product aside from the ones that beforeEach already created. I wonder if there's a more efficient way of cleaning them up without deleting all orders and products, and a way to set them up without creating duplicate orders/products in some of the new tests. I'm afraid that deleting all of them after each test would interfere with other tests when our test suite gained the ability to be run in parallel in the future.

Makes sense, if we expect the tests to be run in parallel, then we need to reduce potential risks that might compromise the test suite stability. I usually try to apply the [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) principle, but parallel execution looks like a feature that will be surely added at some point, so I think it's good that we start preparing for it now in order to reduce the amount of technical debt later.

I'll check if we have any method like `withRestApi.deleteOrder( orderId )` or similar (and the same for products), that I could use for this purpose. If there isn't any, I will create them.

Thank you very much, @rodelgc!

rodelgc commented on packages/js/e2e-core-tests/src/specs/merchant/wp-admin-order-edit.test.js:
```diff
@@ -69,14 +77,152 @@ const runEditOrderTest = () => {
+
+			// Verify file download cannot start
+	  		await merchant.verifyCannotDownloadFromBecause( downloadPage, expectedReason );
+		} );
```
In headless mode, this test (`should not allow downloading a product if download attempts are exceeded`) and the next test (`should not allow downloading a product if expiration date is exceeded`) are passing on my end. 

However, when I ran these two tests using `it.only()` in UI mode, the  `download attempts are exceeded` test got stuck on this part, and did not proceed to the `expiration date is exceeded` test:
<img width="1523" alt="Screen Shot 2022-01-21 at 10 29 42 PM" src="https://user-images.githubusercontent.com/4509348/150547665-8567271b-e3dc-4fac-811d-6738a9bc7d97.png">

Do you also experience the same behavior when running these two tests in `test:e2e-dev` mode?

I have a feeling that this is because the `WordPress > Error` tab that was switched to [here](https://github.com/woocommerce/woocommerce/pull/31650/files#diff-131231274174823411e708b8712c4a010dfb09122a4bc6e4fc7a8a45ec443c2dR288) was not closed after verifying the error message.

Maybe a combination of `browser.pages()` and `page.close()` could help, but I'm not 100% sure as I haven't personally encountered the need for them so far.

alopezari replied:
Oh nice one, I didn't spot that, thanks! I could reproduce it on my end.

```
await merchant.verifyCannotDownloadFromBecause( downloadPage, expectedReason );
```

We are passing the download page to that function, so we just need to close the page inside and that would solve the problem:
```
verifyCannotDownloadFromBecause: async ( page, reason ) => {
	// Select download page tab
	await page.bringToFront();

	// Verify error in download page
	await expect( page.title() ).resolves.toMatch( 'WordPress › Error' );
	await expect( page ).toMatchElement( 'div.wp-die-message', {
		text: reason
	} );

	// Close tab
	await page.close();
}
```
I checked it in both headless and browser modes and it passed.

Thanks again!

rodelgc replied:
> We are passing the download page to that function, so we just need to close the page inside

Ah right! Yep I believe that would solve it 👍

alopezari committed:
- Fixed minor issue from wp-admin-order-edit.test.js when running tests in browser mode.
- Changed those tests so that products and orders are deleted individually after every test run.
- Added deleteOrder() and deleteProduct() functions to withRestApi.

Commit:
Merge branch 'trunk' into alopezari_downloadable-products

rodelgc reviewed (approved):
Looking good. The changes to the 2 tests (`update-product-settings.test.js` and `order-edit.test.js`) are now passing on my end both in headless and `e2e-dev` mode after the change requests were addressed.

I went ahead and [resolved](https://github.com/woocommerce/woocommerce/pull/31650/commits/1672539690abb1fcc6bf7e5cd6ec83ce64b84a90) the merge conflicts in `packages/js/e2e-core-tests/CHANGELOG.md` since the fix was quite simple anyway.

Nice work on this @alopezari!
